app/src/behaviors/behavior_tap_dance.c

```diff
@@ -0,0 +1,347 @@
+/*
+ * Copyright (c) 2022 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_behavior_tap_dance
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <logging/log.h>
+#include <zmk/behavior.h>
+
+#include <zmk/matrix.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/position_state_changed.h>
+#include <zmk/events/keycode_state_changed.h>
+#include <zmk/hid.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+#define ZMK_BHV_TAP_DANCE_MAX_HELD 10
+
+#define ZMK_BHV_TAP_DANCE_POSITION_FREE ULONG_MAX
+
+struct behavior_tap_dance_config {
+    uint32_t tapping_term_ms;
+    int behavior_count;
+    struct zmk_behavior_binding *behaviors;
+};
+
+struct active_tap_dance {
+    // Tap Dance Data
+    int counter;
+    uint32_t position;
+    uint32_t param1;
+    uint32_t param2;
+    bool is_pressed;
+    const struct behavior_tap_dance_config *config;
+
+    // Timer Data
+    bool timer_started;
+    bool timer_cancelled;
+    bool tap_dance_decided;
+    int64_t release_at;
+    struct k_delayed_work release_timer;
+};
+
+// Array to store all active_tap_dances
+struct active_tap_dance active_tap_dances[ZMK_BHV_TAP_DANCE_MAX_HELD] = {};
+
+// *find_tap_dance
+
+// Searches through array of all active tap-dances (active_tap_dances)
+// by checking if a tap-dance in the array exists with the same position
+// as the active keybind.
+// Returns pointer to the tap-dance found at the given position.
+static struct active_tap_dance *find_tap_dance(uint32_t position) {
+    for (int i = 0; i < ZMK_BHV_TAP_DANCE_MAX_HELD; i++) {
+        if (active_tap_dances[i].position == position && !active_tap_dances[i].timer_cancelled) {
+            return &active_tap_dances[i];
+        }
+    }
+    return NULL;
+}
+
+// *new_tap_dance
+
+// Searches through array of all active tap-dances (active_tap_dances)
+// for first element with FREE position. Set its variables to default
+// values as to overwrite any settings from previous tap-dances.
+// Returns pointer to reset active_tap_dance created by this function.
+static struct active_tap_dance *new_tap_dance(uint32_t position,
+                                              const struct behavior_tap_dance_config *config) {
```

Comment:
Since this could fail if we hit the max held, perhaps better to be:

```suggestion
static int new_tap_dance(uint32_t position,
                                              const struct behavior_tap_dance_config *config, struct active_tap_dance **tap_dance) {
```

And update the pointer w/ the selected active tap dance if we have an open slot, return 0, otherwise return `-ENOMEM`.